zaps/0002-encrypt-push-notifications.md

```diff
@@ -0,0 +1,182 @@
+    encrypted_push_registration: Ciphertext # A PushRegistration, encrypted with bouncer_key.
+    push_account_id: RandomID[32]  # Generated by client; need not be unique.
+    bouncer_device_id: RandomID[32] | None  # Generated by bouncer as its unique ID for this record. Null until bouncer has confirmed the registration.
+    push_key: SymmetricKey # Shared key between client and server.
+  ```
+- On the push bouncer, add a new table (parallel to `RemotePushDeviceToken`):
+  ```
+  class zilencer.models.RemotePushDevice:
+    device_id: RandomID[32]  # Unique key for this record, generated by bouncer.
+    realm: ForeignKey[RemoteRealm]
+    token_kind: "apns" | "fcm"
+    token: str  # Provided by client.
+    push_account_id: RandomID[32] # Generated by client; need not be unique.
+    expired_time: datetime | None
+  ```
+
+- We probably also need a `zerver.models.push_notifications.PushDevice`
+  for Zulip Cloud use with no third-party bouncer? Or we could make Cloud use a bouncer.
+
+- Additionally, the push bouncer has an asymmetric key `bouncer_key`,
+  of which the public part is baked into clients.
+
+## Operation
+
+### To register a new push device for a Zulip account
+
+1. The client checks the [`/api/v1/server_settings`](https://zulip.com/api/get-server-settings)
+   to see whether the server advertises the `encrypted_push_notifications`
+   capability (or has a new enough feature level, details TBD).
+
+   If it does not, it uses the legacy push notifications registration flow.
+   (Note: We’ll likely want settings to require E2E encryption; that can come
+   after the feature is complete).
+
+1. The client generates a new RandomID `push_account_id`. It builds a JSON
+   object `push_registration` of type `PushRegistration`, and encrypts it to
+   `bouncer_key` as `encrypted_push_registration`.
+
+1. The client makes an API call to the server (authenticated with the user’s
+   API key) with arguments `push_account_id` and `encrypted_push_registration`.
+
+   The server generates a symmetric key `push_key` and creates a `PushDevice`
+   instance. The server replies to the client with `push_key`, and asynchronously
+   initiates the next step.
+
+   - Next step asynchronous: It ensures that the bouncer service 500ing will just
+     cause queueing of those requests for the local server, the mobile client
+     doesn’t need to deal with retries here. Handling of 500 is discussed below.
+
+   - TBD: Optionally, the client could include some randomness to be mixed into
+     `push_key`. We can decide if this is needed when reviewing the exact crypto
+     primitives we end up using.
+
+1. The server makes an API call to the push bouncer (authenticated with the
+   server’s API key) with arguments `user_id`, `user_uuid`, `realm_uuid`,
+   `push_account_id`, and `encrypted_push_registration`.
+
+   The push bouncer:
+   1. Decrypts `push_registration`.
+
+   1. Checks the `timestamp` for liveness, probably 60s is a good age limit.
```

Comment:
Could be, but more like a day was the design in the previous discussion a couple of years ago.

Reply:
Copying from a thread there:
> [Having this liveness check] prevents replays, whether as part of an attack or due to retries in the presence of connection failures.
>
> Because we also check for duplicates, this can be a long limit; the only reason to keep it short is so that we can eventually GC records that we learn from APNs/FCM are stale or invalid, because we don’t want to do so while there could still be a valid old request to add the same record in flight.

Reply:
OK, maybe 24 hours is a fine default then? We might want to name this value, say `PUSH_REGISTRATION_LIVENESS_TIMEOUT = 60 * 60 * 24`, with the intent that we'd give it a nice comment in the source code.

Reply:
```diff
-   1. Checks the `timestamp` for liveness, probably 60s is a good age limit.
+   1. Check the `timestamp` for liveness, `PUSH_REGISTRATION_LIVENESS_TIMEOUT`
+      which equals `24 * 60 * 60` seconds is a good age limit because
+      a day is longer than any minor outage.
+
       A coded `JsonableError` should be used if the timestamp is stale.
```